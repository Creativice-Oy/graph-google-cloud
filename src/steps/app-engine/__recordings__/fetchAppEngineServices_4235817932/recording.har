{
  "log": {
    "_recordingName": "fetchAppEngineServices",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 739,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "739"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 283,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/json",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 565,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 565,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 11 Mar 2021 14:57:50 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 511,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-03-11T14:57:50.653Z",
        "time": 170,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 170
        }
      },
      {
        "_id": "a8077ca10a73994713b7a0d586480166",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/4.4.3 gl-node/14.15.4 auth/6.0.5"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/4.4.3 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "appengine.googleapis.com"
            }
          ],
          "headersSize": 538,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://appengine.googleapis.com/v1/apps/j1-gc-integration-dev-v2"
        },
        "response": {
          "bodySize": 637,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 637,
            "text": "{\"name\":\"apps/j1-gc-integration-dev-v2\",\"id\":\"j1-gc-integration-dev-v2\",\"authDomain\":\"gmail.com\",\"locationId\":\"us-central\",\"codeBucket\":\"staging.j1-gc-integration-dev-v2.appspot.com\",\"servingStatus\":\"SERVING\",\"defaultHostname\":\"j1-gc-integration-dev-v2.uc.r.appspot.com\",\"defaultBucket\":\"j1-gc-integration-dev-v2.appspot.com\",\"gcrDomain\":\"us.gcr.io\",\"databaseType\":\"CLOUD_DATASTORE_COMPATIBILITY\",\"featureSettings\":{\"splitHealthChecks\":true,\"useContainerOptimizedOs\":true}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 11 Mar 2021 14:57:51 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 488,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-03-11T14:57:50.835Z",
        "time": 455,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 455
        }
      },
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 1,
        "cache": {},
        "request": {
          "bodySize": 739,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "739"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 283,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/json",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 552,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 552,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 11 Mar 2021 14:58:50 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 511,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-03-11T14:58:49.917Z",
        "time": 169,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 169
        }
      },
      {
        "_id": "83742efc2d14453e4066991a3f93e347",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/4.4.3 gl-node/14.15.4 auth/6.0.5"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/4.4.3 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "appengine.googleapis.com"
            }
          ],
          "headersSize": 547,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://appengine.googleapis.com/v1/apps/j1-gc-integration-dev-v2/services"
        },
        "response": {
          "bodySize": 311,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 311,
            "text": "{\"services\":[{\"name\":\"apps/j1-gc-integration-dev-v2/services/default\",\"id\":\"default\",\"split\":{\"allocations\":{\"20210309t141022\":1}}}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Thu, 11 Mar 2021 14:58:50 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 488,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-03-11T14:58:50.091Z",
        "time": 461,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 461
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
