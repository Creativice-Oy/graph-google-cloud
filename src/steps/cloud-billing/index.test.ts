import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { integrationConfig } from '../../../test/config';
import { setupGoogleCloudRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../types';
import { buildProjectBudgetRelationships, fetchBillingBudget } from '.';
import {
  fetchResourceManagerProject,
  PROJECT_ENTITY_TYPE,
} from '../resource-manager/index';
import {
  ENTITY_TYPE_BILLING_BUDGET,
  RELATIONSHIP_TYPE_PROJECT_HAS_BUDGET,
} from './constants';

describe('#fetchBillingBudget', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchBillingBudget',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });

    await fetchBillingBudget(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_BILLING_BUDGET,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Entity'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_billing_budget' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          projects: { type: 'array', items: { type: 'string' } },
          specifiedAmoutCurrencyCode: { type: 'string' },
          specifiedAmoutUnits: { type: 'string' },
          specifiedAmoutNanos: { type: 'string' },
          pubsubTopic: { type: 'string' },
          schemaVersion: { type: 'string' },
          monitoringNotificationChannels: { type: 'string' },
          disableDefaultIamRecipients: { type: 'string' },
          etag: { type: 'string' },
        },
      },
    });
  });
});

describe('#buildProjectBudgetRelationships', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'buildProjectBudgetRelationships',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: {
        ...integrationConfig,
        serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
          'j1-gc-integration-dev-v2',
          'j1-gc-integration-dev-v3',
        ),
        serviceAccountKeyConfig: {
          ...integrationConfig.serviceAccountKeyConfig,
          project_id: 'j1-gc-integration-dev-v3',
        },
      },
    });

    await fetchBillingBudget(context);
    await fetchResourceManagerProject(context);
    await buildProjectBudgetRelationships(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === PROJECT_ENTITY_TYPE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_project' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          projectId: { type: 'string' },
          name: { type: 'string' },
          displayName: { type: 'string' },
          parent: { type: 'string' },
          lifecycleState: { type: 'string' },
          createdOn: { type: 'number' },
          updatedOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_BILLING_BUDGET,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Entity'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_billing_budget' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          projects: { type: 'array', items: { type: 'string' } },
          specifiedAmoutCurrencyCode: { type: 'string' },
          specifiedAmoutUnits: { type: 'string' },
          specifiedAmoutNanos: { type: 'string' },
          pubsubTopic: { type: 'string' },
          schemaVersion: { type: 'string' },
          monitoringNotificationChannels: { type: 'string' },
          disableDefaultIamRecipients: { type: 'string' },
          etag: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_PROJECT_HAS_BUDGET,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'google_cloud_project_has_billing_budget',
          },
        },
      },
    });
  });
});
