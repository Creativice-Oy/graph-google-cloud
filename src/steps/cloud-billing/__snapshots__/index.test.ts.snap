// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildProjectBudgetRelationships should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
              "projects": Array [
                "projects/167984947943",
              ],
            },
            "displayName": "budget-example-01",
            "etag": "1627655637440882",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-01",
      "etag": "1627655637440882",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "projects": Array [
        "projects/167984947943",
      ],
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
            },
            "displayName": "budget-example-multi-project",
            "etag": "1627656753474895",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-multi-project",
      "etag": "1627656753474895",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
      "projects": undefined,
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
              "projects": Array [
                "projects/538466736102",
                "projects/167984947943",
              ],
            },
            "displayName": "budget-example-some-projects",
            "etag": "1627656774053178",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-some-projects",
      "etag": "1627656774053178",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "projects": Array [
        "projects/538466736102",
        "projects/167984947943",
      ],
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "projects/167984947943",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createTime": "2021-05-31T15:15:56.125Z",
            "displayName": "j1-gc-integration-dev-v3",
            "etag": "CzUfI2XTWvMJgVGlic9LWA==",
            "name": "projects/167984947943",
            "parent": "folders/306329820365",
            "projectId": "j1-gc-integration-dev-v3",
            "state": "ACTIVE",
            "updateTime": "2021-05-31T15:15:58.148Z",
          },
        },
      ],
      "_type": "google_cloud_project",
      "createdOn": 1622474156125,
      "displayName": "j1-gc-integration-dev-v3",
      "id": "j1-gc-integration-dev-v3",
      "lifecycleState": "ACTIVE",
      "name": "projects/167984947943",
      "parent": "folders/306329820365",
      "projectId": "j1-gc-integration-dev-v3",
      "updatedOn": 1622474158148,
      "webLink": "https://console.cloud.google.com/home/dashboard?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "_toEntityKey": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "_type": "google_cloud_project_has_billing_budget",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_key": "projects/538466736102|has|billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "skipTargetCreation": true,
        "sourceEntityKey": "projects/538466736102",
        "targetEntity": Object {
          "_class": Array [
            "Entity",
          ],
          "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
          "_rawData": undefined,
          "_type": "google_billing_budget",
          "createdOn": undefined,
          "disableDefaultIamRecipients": undefined,
          "displayName": "budget-example-some-projects",
          "etag": "1627656774053178",
          "monitoringNotificationChannels": undefined,
          "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
          "projects": Array [
            "projects/538466736102",
            "projects/167984947943",
          ],
          "pubsubTopic": undefined,
          "schemaVersion": undefined,
          "specifiedAmoutCurrencyCode": "USD",
          "specifiedAmoutNanos": undefined,
          "specifiedAmoutUnits": "1000",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "google_cloud_project_has_billing_budget",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "projects/167984947943",
      "_key": "projects/167984947943|has|billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "_toEntityKey": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "_type": "google_cloud_project_has_billing_budget",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "google_billing_budget",
    "google_cloud_project",
    "google_cloud_project_has_billing_budget",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 3,
}
`;

exports[`#fetchBillingBudget should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
              "projects": Array [
                "projects/167984947943",
              ],
            },
            "displayName": "budget-example-01",
            "etag": "1627655637440882",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-01",
      "etag": "1627655637440882",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/459b5dfa-1c2b-47af-8396-e3fd5e32a834",
      "projects": Array [
        "projects/167984947943",
      ],
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
            },
            "displayName": "budget-example-multi-project",
            "etag": "1627656753474895",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-multi-project",
      "etag": "1627656753474895",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/7fa6803c-0e1a-415e-ab30-69d953298d11",
      "projects": undefined,
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
    Object {
      "_class": Array [
        "Entity",
      ],
      "_key": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "amount": Object {
              "specifiedAmount": Object {
                "currencyCode": "USD",
                "units": "1000",
              },
            },
            "budgetFilter": Object {
              "calendarPeriod": "MONTH",
              "creditTypesTreatment": "INCLUDE_ALL_CREDITS",
              "projects": Array [
                "projects/538466736102",
                "projects/167984947943",
              ],
            },
            "displayName": "budget-example-some-projects",
            "etag": "1627656774053178",
            "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
            "notificationsRule": Object {},
            "thresholdRules": Array [
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.5,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 0.9,
              },
              Object {
                "spendBasis": "CURRENT_SPEND",
                "thresholdPercent": 1,
              },
            ],
          },
        },
      ],
      "_type": "google_billing_budget",
      "createdOn": undefined,
      "disableDefaultIamRecipients": undefined,
      "displayName": "budget-example-some-projects",
      "etag": "1627656774053178",
      "monitoringNotificationChannels": undefined,
      "name": "billingAccounts/01694E-12D60B-BEBCE5/budgets/c7f7c572-2c36-4d77-bd73-a5090d150177",
      "projects": Array [
        "projects/538466736102",
        "projects/167984947943",
      ],
      "pubsubTopic": undefined,
      "schemaVersion": undefined,
      "specifiedAmoutCurrencyCode": "USD",
      "specifiedAmoutNanos": undefined,
      "specifiedAmoutUnits": "1000",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "google_billing_budget",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 0,
}
`;
